Readme.md
kanjivg2svg
===========

This Ruby 1.9 script takes stroke order data from the [KanjiVG](http://kanjivg.tagaini.net/) project and outputs SVG files with special formatting.

Usage
-----

    $ mkdir svgs
    $ ruby kanjivg2svg.rb path/to/kanjivg.xml [frames|animated|numbers]

You can change the output type by setting the second argument. If not set it will default to 'frames'. The animated and numbers are less perfected compared to the frames output.

In this repo I've included svg files generated with the 'frames' option.

License
-------

By Kim Ahlström <kim.ahlstrom@gmail.com>

[Creative Commons Attribution-Share Alike 3.0](http://creativecommons.org/licenses/by-sa/3.0/)

KanjiVG
-------

KanjiVG is copyright (c) 2009/2010 Ulrich Apel and released under the Creative Commons Attribution-Share Alike 3.0


Changelog.md
March 27, 2011
--------------

These files were modified by 10ecac732a0bb3555ca95a218708c870306cb5ae and 14245e982fec627e3ff89744ccb33551f4b3b624.

#	modified:   101_frames.svg
#	modified:   12362_frames.svg
#	modified:   12399_frames.svg
#	modified:   12401_frames.svg
#	modified:   12408_frames.svg
#	modified:   12409_frames.svg
#	modified:   12410_frames.svg
#	modified:   12434_frames.svg
#	modified:   12452_frames.svg
#	modified:   12454_frames.svg
#	modified:   12526_frames.svg
#	modified:   12532_frames.svg
#	modified:   13505_frames.svg
#	modified:   20098_frames.svg
#	modified:   20559_frames.svg
#	modified:   20717_frames.svg
#	modified:   20842_frames.svg
#	modified:   20876_frames.svg
#	modified:   20883_frames.svg
#	modified:   20960_frames.svg
#	modified:   20973_frames.svg
#	modified:   20976_frames.svg
#	modified:   21208_frames.svg
#	modified:   21241_frames.svg
#	modified:   21353_frames.svg
#	modified:   21371_frames.svg
#	modified:   21471_frames.svg
#	modified:   21836_frames.svg
#	modified:   21972_frames.svg
#	modified:   22040_frames.svg
#	modified:   22063_frames.svg
#	modified:   22196_frames.svg
#	modified:   22343_frames.svg
#	modified:   22518_frames.svg
#	modified:   22570_frames.svg
#	modified:   22797_frames.svg
#	modified:   23094_frames.svg
#	modified:   23398_frames.svg
#	modified:   23526_frames.svg
#	modified:   23557_frames.svg
#	modified:   23692_frames.svg
#	modified:   24090_frames.svg
#	modified:   24931_frames.svg
#	modified:   25120_frames.svg
#	modified:   25233_frames.svg
#	modified:   25499_frames.svg
#	modified:   25826_frames.svg
#	modified:   26039_frames.svg
#	modified:   26047_frames.svg
#	modified:   26114_frames.svg
#	modified:   26332_frames.svg
#	modified:   26791_frames.svg
#	modified:   26922_frames.svg
#	modified:   27075_frames.svg
#	modified:   27148_frames.svg
#	modified:   27310_frames.svg
#	modified:   27663_frames.svg
#	modified:   27665_frames.svg
#	modified:   27733_frames.svg
#	modified:   28079_frames.svg
#	modified:   28149_frames.svg
#	modified:   28189_frames.svg
#	modified:   28399_frames.svg
#	modified:   28809_frames.svg
#	modified:   29014_frames.svg
#	modified:   30008_frames.svg
#	modified:   30032_frames.svg
#	modified:   30476_frames.svg
#	modified:   30494_frames.svg
#	modified:   30501_frames.svg
#	modified:   30502_frames.svg
#	modified:   30669_frames.svg
#	modified:   30789_frames.svg
#	modified:   31047_frames.svg
#	modified:   31163_frames.svg
#	modified:   31492_frames.svg
#	modified:   31786_frames.svg
#	modified:   32025_frames.svg
#	modified:   32026_frames.svg
#	modified:   32768_frames.svg
#	modified:   32862_frames.svg
#	modified:   32886_frames.svg
#	modified:   33081_frames.svg
#	modified:   33323_frames.svg
#	modified:   33914_frames.svg
#	modified:   34659_frames.svg
#	modified:   34914_frames.svg
#	modified:   35219_frames.svg
#	modified:   35241_frames.svg
#	modified:   35264_frames.svg
#	modified:   35624_frames.svg
#	modified:   36011_frames.svg
#	modified:   36795_frames.svg
#	modified:   36802_frames.svg
#	modified:   36804_frames.svg
#	modified:   36817_frames.svg
#	modified:   36864_frames.svg
#	modified:   36865_frames.svg
#	modified:   36880_frames.svg
#	modified:   36883_frames.svg
#	modified:   36889_frames.svg
#	modified:   36895_frames.svg
#	modified:   36896_frames.svg
#	modified:   36898_frames.svg
#	modified:   36924_frames.svg
#	modified:   36929_frames.svg
#	modified:   36933_frames.svg
#	modified:   36935_frames.svg
#	modified:   36938_frames.svg
#	modified:   36939_frames.svg
#	modified:   36941_frames.svg
#	modified:   36942_frames.svg
#	modified:   36946_frames.svg
#	modified:   36949_frames.svg
#	modified:   36961_frames.svg
#	modified:   36969_frames.svg
#	modified:   36986_frames.svg
#	modified:   37406_frames.svg
#	modified:   38017_frames.svg
#	modified:   38520_frames.svg
#	modified:   39214_frames.svg
#	modified:   39768_frames.svg
#	modified:   40199_frames.svg
#	modified:   51_frames.svg
#	modified:   68_frames.svg


convert.rb
# Encoding: UTF-8

require 'rubygems'
require 'nokogiri'
require 'pp'

class Importer
  class KanjiVG
    
    WIDTH = 109 # 109 per character
    HEIGHT = 109 # 109 per character
    SVG_HEAD = "<svg width=\"__WIDTH__px\" height=\"#{HEIGHT}px\" viewBox=\"0 0 __VIEW_WIDTH__px #{HEIGHT}px\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" xml:space=\"preserve\" version=\"1.1\"  baseProfile=\"full\">"
    SVG_FOOT = '</svg>'
    TEXT_STYLE = 'fill:#FF2A00;font-family:Helvetica;font-weight:normal;font-size:14;stroke-width:0'
    PATH_STYLE = 'fill:none;stroke:black;stroke-width:3'
    INACTIVE_PATH_STYLE = 'fill:none;stroke:#999;stroke-width:3'
    LINE_STYLE = 'stroke:#ddd;stroke-width:2'
    DASHED_LINE_STYLE = 'stroke:#ddd;stroke-width:2;stroke-dasharray:3 3'
    ENTRY_NAME = 'kanji'
    COORD_RE = %r{(?ix:\d+ (?:\.\d+)?)}
    
    def initialize(doc, output_dir, type = :numbers, verbose = false)
      @output_dir = output_dir
      @type = type
      @verbose = verbose
      processed = 0

      puts "Starting the conversion @ #{Time.now} ..." if @verbose
      
      # Don't want Nokogiri to read in the entire document at once
      # So doing it entry by entry
      tmp = ""
      begin
        while (line = doc.readline)
          if line =~ %r{<#{ENTRY_NAME}}
            tmp = line
          elsif line =~ %r{</#{ENTRY_NAME}>}
            tmp << line
            noko = Nokogiri::XML(tmp)
            parse(noko)
            
            processed += 1
            if processed % 200 == 0
              puts "Processed #{processed} @ #{Time.now}" if @verbose
            end
          else
            tmp << line
          end
        end
      rescue EOFError
        doc.close
      end
    end
    
    private
    
    def parse(doc)
      doc.css(ENTRY_NAME).each do |entry|
        codepoint = entry['id']
        svg = File.open("#{@output_dir}/U+#{codepoint}_#{@type}.svg", File::RDWR|File::TRUNC|File::CREAT)
        stroke_count = 0
        stroke_total = entry.css('stroke[path]').length
        paths = []
        
        if @type == :frames
          width = (WIDTH * stroke_total)# + (2 * (stroke_total - 1))
          view_width = width
        else
          width = WIDTH * 1
        end
        header = SVG_HEAD.gsub('__WIDTH__', width.to_s)
        header = header.gsub('__VIEW_WIDTH__', view_width.to_s)
        svg << "#{header}\n"
        
        if @type == :frames
          # Outer box
          top = 1; left = 1; bottom = HEIGHT - 1; right = width - 1
          svg << line(left, top, right, top, LINE_STYLE) # top
          svg << line(left, top, left, bottom, LINE_STYLE) # left
          svg << line(left, bottom, right, bottom, LINE_STYLE) # bottom
          svg << line(right, top, right, bottom, LINE_STYLE) # right
          
          (1 .. stroke_total - 1).each do |i|
            svg << line(WIDTH * i, top, WIDTH * i, bottom, LINE_STYLE)
          end
          
          # Inner guides
          svg << line(left, (HEIGHT/2), right, (HEIGHT/2), DASHED_LINE_STYLE)
          
          (1 .. stroke_total).each do |i|
            svg << line((WIDTH/2)+(WIDTH*(i-1)+1), top, (WIDTH/2)+(WIDTH*(i-1)+1), bottom, DASHED_LINE_STYLE)
          end
        end
        
        entry.css('stroke[path]').each do |stroke|
          paths << stroke['path']
          stroke_count += 1
          
          case @type
          when :animated
            svg << "<path d=\"#{stroke['path']}\" style=\"#{PATH_STYLE};opacity:0\">\n"
            svg << "  <animate attributeType=\"CSS\" attributeName=\"opacity\" from=\"0\" to=\"1\" begin=\"#{stroke_count-1}s\" dur=\"1s\" repeatCount=\"0\" fill=\"freeze\" />\n"
            svg << "</path>\n"
          when :numbers
            x, y = move_text_relative_to_path(stroke['path'])
            svg << "<text x=\"#{x}\" y=\"#{y}\" style=\"#{TEXT_STYLE}\">#{stroke_count}</text>\n"
            svg << "<path d=\"#{stroke['path']}\" style=\"#{PATH_STYLE}\" />\n"
          when :frames
            md = %r{^[LMT] (#{COORD_RE}) , (#{COORD_RE})}ix.match(paths.last)
            path_start_x = md[1].to_f
            path_start_y = md[2].to_f
            path_start_x += WIDTH * (stroke_count - 1)
            
            paths.each_with_index do |path, i|
              last = ((stroke_count - 1) == i)
              delta = last ? WIDTH * (stroke_count - 1) : WIDTH
              
              path.gsub!(%r{([LMT]) (#{COORD_RE})}x) do |m|
                letter = $1
                x  = $2.to_f
                x += delta
                "#{letter}#{x}"
              end
              path.gsub!(%r{(S) (#{COORD_RE}) , (#{COORD_RE}) , (#{COORD_RE})}x) do |m|
                letter = $1
                x1  = $2.to_f
                x1 += delta
                x2  = $4.to_f
                x2 += delta
                "#{letter},#{x1},#{$3},#{x2}"
              end
              path.gsub!(%r{(C) (#{COORD_RE}) , (#{COORD_RE}) , (#{COORD_RE}) , (#{COORD_RE}) , (#{COORD_RE})}x) do |m|
                letter  = $1
                x1  = $2.to_f
                x1 += delta
                x2  = $4.to_f
                x2 += delta
                x3  = $6.to_f
                x3 += delta
                "#{letter}#{x1},#{$3},#{x2},#{$5},#{x3}"
              end
              
              svg << "<path d=\"#{path}\" style=\"#{last ? PATH_STYLE : INACTIVE_PATH_STYLE}\" />\n"
            end
            
            # Stroke start
            svg << "<circle cx=\"#{path_start_x}\" cy=\"#{path_start_y}\" r=\"5\" stroke-width=\"0\" fill=\"#FF2A00\" opacity=\"0.7\" />"
            svg << "\n"
          end
        end

        svg << SVG_FOOT
        svg.close
      end
    end
    
    # TODO: make this shit really smart
    def move_text_relative_to_path(path)
      md = %r{^M (#{COORD_RE}) , (#{COORD_RE})}ix.match(path)
      path_start_x = md[1].to_f
      path_start_y = md[2].to_f
      
      text_x = path_start_x
      text_y = path_start_y
      
      [text_x, text_y]
    end
    
    def line(x1, y1, x2, y2, style)
      "<line x1=\"#{x1}\" y1=\"#{y1}\" x2=\"#{x2}\" y2=\"#{y2}\" style=\"#{style}\" />\n"
    end
    
  end
end

file = ARGV[0]
Importer::KanjiVG.new(File.open(file), File.expand_path('../svgs',  __FILE__), :frames, true)


kanjivg2svg.rb
# Encoding: UTF-8
# By: Kim Ahlström <kim.ahlstrom@gmail.com>
# License: Creative Commons Attribution-Share Alike 3.0 - http://creativecommons.org/licenses/by-sa/3.0/
# KanjiVG is copyright (c) 2009/2010 Ulrich Apel and released under the Creative Commons Attribution-Share Alike 3.0

require 'rubygems'
require 'nokogiri'
require 'pp'

class Importer
  class KanjiVG
    
    WIDTH = 109 # 109 per character
    HEIGHT = 109 # 109 per character
    SVG_HEAD = "<svg width=\"__WIDTH__px\" height=\"#{HEIGHT}px\" viewBox=\"0 0 __VIEW_WIDTH__px #{HEIGHT}px\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" xml:space=\"preserve\" version=\"1.1\"  baseProfile=\"full\">"
    SVG_FOOT = '</svg>'
    TEXT_STYLE = 'fill:#FF2A00;font-family:Helvetica;font-weight:normal;font-size:14;stroke-width:0'
    PATH_STYLE = 'fill:none;stroke:black;stroke-width:3'
    INACTIVE_PATH_STYLE = 'fill:none;stroke:#999;stroke-width:3'
    LINE_STYLE = 'stroke:#ddd;stroke-width:2'
    DASHED_LINE_STYLE = 'stroke:#ddd;stroke-width:2;stroke-dasharray:3 3'
    ENTRY_NAME = 'kanji'
    COORD_RE = %r{(?ix:\d+ (?:\.\d+)?)}
    
    def initialize(doc, output_dir, type = :numbers, verbose = false)
      @output_dir = output_dir
      @type = type
      @verbose = verbose
      processed = 0

      puts "Starting the conversion @ #{Time.now} ..." if @verbose
      
      # Don't want Nokogiri to read in the entire document at once
      # So doing it entry by entry
      tmp = ""
      begin
        while (line = doc.readline)
          if line =~ %r{<#{ENTRY_NAME}}
            tmp = line
          elsif line =~ %r{</#{ENTRY_NAME}>}
            tmp << line
            noko = Nokogiri::XML(tmp)
            parse(noko)
            
            processed += 1
            if processed % 200 == 0
              puts "Processed #{processed} @ #{Time.now}" if @verbose
            end
          else
            tmp << line
          end
        end
      rescue EOFError
        doc.close
      end
    end
    
    private
    
    def parse(doc)
      doc.css(ENTRY_NAME).each do |entry|
        codepoint = entry['midashi'].codepoints.first
        svg = File.open("#{@output_dir}/#{codepoint}_#{@type}.svg", File::RDWR|File::TRUNC|File::CREAT)
        stroke_count = 0
        stroke_total = entry.css('stroke[path]').length
        paths = []
        
        # Generate the header
        if @type == :frames
          width = (WIDTH * stroke_total)# + (2 * (stroke_total - 1))
          view_width = width
        else
          width = WIDTH * 1
        end
        header = SVG_HEAD.gsub('__WIDTH__', width.to_s)
        header = header.gsub('__VIEW_WIDTH__', view_width.to_s)
        svg << "#{header}\n"
        
        # Guide lines
        if @type == :frames
          # Outer box
          top = 1; left = 1; bottom = HEIGHT - 1; right = width - 1
          svg << line(left, top, right, top, LINE_STYLE) # top
          svg << line(left, top, left, bottom, LINE_STYLE) # left
          svg << line(left, bottom, right, bottom, LINE_STYLE) # bottom
          svg << line(right, top, right, bottom, LINE_STYLE) # right
          
          (1 .. stroke_total - 1).each do |i|
            svg << line(WIDTH * i, top, WIDTH * i, bottom, LINE_STYLE)
          end
          
          # Inner guides
          svg << line(left, (HEIGHT/2), right, (HEIGHT/2), DASHED_LINE_STYLE)
          
          (1 .. stroke_total).each do |i|
            svg << line((WIDTH/2)+(WIDTH*(i-1)+1), top, (WIDTH/2)+(WIDTH*(i-1)+1), bottom, DASHED_LINE_STYLE)
          end
        end
        
        # Draw the strokes
        entry.css('stroke[path]').each do |stroke|
          paths << stroke['path']
          stroke_count += 1
          
          case @type
          when :animated
            svg << "<path d=\"#{stroke['path']}\" style=\"#{PATH_STYLE};opacity:0\">\n"
            svg << "  <animate attributeType=\"CSS\" attributeName=\"opacity\" from=\"0\" to=\"1\" begin=\"#{stroke_count-1}s\" dur=\"1s\" repeatCount=\"0\" fill=\"freeze\" />\n"
            svg << "</path>\n"
          when :numbers
            x, y = move_text_relative_to_path(stroke['path'])
            svg << "<text x=\"#{x}\" y=\"#{y}\" style=\"#{TEXT_STYLE}\">#{stroke_count}</text>\n"
            svg << "<path d=\"#{stroke['path']}\" style=\"#{PATH_STYLE}\" />\n"
          when :frames
            md = %r{^[LMT] (#{COORD_RE}) , (#{COORD_RE})}ix.match(paths.last)
            path_start_x = md[1].to_f
            path_start_y = md[2].to_f
            path_start_x += WIDTH * (stroke_count - 1)
            
            paths.each_with_index do |path, i|
              last = ((stroke_count - 1) == i)
              delta = last ? WIDTH * (stroke_count - 1) : WIDTH
              
              # Move strokes relative to the frame
              path.gsub!(%r{([LMTm]) (#{COORD_RE})}x) do |m|
                letter = $1
                x  = $2.to_f
                x += delta
                "#{letter}#{x}"
              end
              path.gsub!(%r{(S) (#{COORD_RE}) , (#{COORD_RE}) , (#{COORD_RE})}x) do |m|
                letter = $1
                x1  = $2.to_f
                x1 += delta
                x2  = $4.to_f
                x2 += delta
                "#{letter}#{x1},#{$3},#{x2}"
              end
              path.gsub!(%r{(C) (#{COORD_RE}) , (#{COORD_RE}) , (#{COORD_RE}) , (#{COORD_RE}) , (#{COORD_RE})}x) do |m|
                letter  = $1
                x1  = $2.to_f
                x1 += delta
                x2  = $4.to_f
                x2 += delta
                x3  = $6.to_f
                x3 += delta
                "#{letter}#{x1},#{$3},#{x2},#{$5},#{x3}"
              end
              
              svg << "<path d=\"#{path}\" style=\"#{last ? PATH_STYLE : INACTIVE_PATH_STYLE}\" />\n"
            end
            
            # Put a circle at the stroke start
            svg << "<circle cx=\"#{path_start_x}\" cy=\"#{path_start_y}\" r=\"5\" stroke-width=\"0\" fill=\"#FF2A00\" opacity=\"0.7\" />"
            svg << "\n"
          end
        end

        svg << SVG_FOOT
        svg.close
      end
    end
    
    # TODO: make this shit really smart
    def move_text_relative_to_path(path)
      md = %r{^M (#{COORD_RE}) , (#{COORD_RE})}ix.match(path)
      path_start_x = md[1].to_f
      path_start_y = md[2].to_f
      
      text_x = path_start_x
      text_y = path_start_y
      
      [text_x, text_y]
    end
    
    def line(x1, y1, x2, y2, style)
      "<line x1=\"#{x1}\" y1=\"#{y1}\" x2=\"#{x2}\" y2=\"#{y2}\" style=\"#{style}\" />\n"
    end
    
  end
end

file = ARGV[0]
type = ARGV[1] || 'frames'
Importer::KanjiVG.new(File.open(file), File.expand_path('../svgs',  __FILE__), type.to_sym, true)
